kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal;
+
+import java.util.*;
+import static java.util.Objects.requireNonNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.delta.kernel.*;
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.expressions.Predicate;
+import io.delta.kernel.types.*;
+
+import io.delta.kernel.internal.actions.*;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.replay.LogReplay;
+import io.delta.kernel.internal.snapshot.*;
+import io.delta.kernel.internal.util.Tuple2;
+import static io.delta.kernel.internal.TransactionImpl.DEFAULT_READ_VERSION;
+import static io.delta.kernel.internal.TransactionImpl.DEFAULT_WRITE_VERSION;
+import static io.delta.kernel.internal.util.VectorUtils.stringArrayValue;
+import static io.delta.kernel.internal.util.VectorUtils.stringStringMapValue;
+
+public class TransactionBuilderImpl implements TransactionBuilder {
+    private static final Logger logger = LoggerFactory.getLogger(TransactionBuilderImpl.class);
+
+    private final long currentTimeMillis = System.currentTimeMillis();
+    private final TableImpl table;
+    private final String engineInfo;
+    private final String operation;
+    private Optional<StructType> schema = Optional.empty();
+    private Optional<Set<String>> partitionColumns = Optional.empty();
+    private Optional<Predicate> readPredicate = Optional.empty();
+    private Optional<Long> readVersion = Optional.empty();
+    private Optional<SetTransaction> transactionId = Optional.empty();
+
+    public TransactionBuilderImpl(TableImpl table, String engineInfo, String operation) {
+        this.table = table;
+        this.engineInfo = engineInfo;
+        this.operation = operation;
+    }
+
+    @Override
+    public TransactionBuilder withSchema(TableClient tableClient, StructType newSchema) {
+        this.schema = Optional.of(newSchema);
+        // TODO: this needs to verified
+        return this;
+    }
+
+    @Override
+    public TransactionBuilder withPartitionColumns(
```

Comment:
worth commenting that this is asserted it happens on a new table in the validate method before build?